### PR TITLE
[WIP] Move to pull-streams

### LINKED ISSUE #67: Update encodings following the changes in go-ipfs
Ref 

- https://github.com/ipfs/go-datastore/pull/39
- https://github.com/ipfs/go-datastore/pull/41
- https://github.com/ipfs/go-ipfs/pull/2903

New paths are `/block/${base-32-std-no-padding-encoded-hash}`

### PR DESCRIPTION
**Pull request by dignifiedquire**
This uses the newly created [interface-pull-blob-store](https://github.com/ipfs/interface-pull-blob-store) with the `fs` and `id` implementations. The interface remains the same, except the renaming of

- `datastore` - > `blockstore`
- `datastore-legacy` - >`datastore`

This also fixes #67 and is as such a breaking change, but can be mitigated using `fs-repo-migrations` for fs nodes.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ sudo: false
 language: node_js
 node_js:
   - 4
-  - 5
+  - stable
```
Comment (diasdavid) [inline .travis.yml]: - [ ] add Nodejs 5 too

File: README.md [hunk 2]
```diff
@@ -173,12 +173,14 @@ Read/write keys inside the repo. This feature will be expanded once
 [IPRS](https://github.com/ipfs/specs/tree/master/records) and
 [KeyChain](https://github.com/ipfs/specs/tree/master/keychain) are finalized and implemented on go-ipfs.
 
-### repo.datastore.read(key, cb(err, buffer))
-### repo.datastore.write(buffer, cb(err, buffer))
+### repo.blockstore.putStream()
+### repo.datastore.getStream(key, extension)
+### repo.datastore.has(key, extension, cb)
+### repo.datastore.delete(key, extension, cb)
```
Comment (diasdavid) [inline README.md]: - [ ] add docs (specially types of the arguments and what they return)

File: src/index.js [hunk 3]
```diff
@@ -1,53 +1,55 @@
...
+      if (needsLocks) {
+        args.push(this.locks)
+      }
+
+      if (needsConfig) {
+        args.push(this.config)
+      }
+
+      return stores[name].setUp.apply(stores[name], args)
+    }
```
Comment (diasdavid) [inline src/index.js]: This is a new step. What is the 'needsLocks' and 'needsConfig'?

File: src/index.js [hunk 4]
```diff
@@ -1,53 +1,55 @@
...
+    assert.equal(typeof repoPath, 'string', 'missing repoPath')
+    assert(options, 'missing options')
+    assert(options.stores, 'missing options.stores')
+
+    this.path = repoPath
+
+    const blobStores = initializeBlobStores(options.stores)
+
+    const setup = (name, needsLocks, needsConfig) => {
+      const args = [repoPath, blobStores[name]]
```
Comment (diasdavid) [inline src/index.js]: Node.js 4 and 5 won't like this, IIRC

File: src/index.js (hunk 3 above)
Comment (diasdavid) [inline src/index.js]: ok got it.

- [ ] make it a options object

File: src/index.js [hunk 5]
```diff
@@ -1,53 +1,55 @@
...
+
+    this.locks = setup('locks')
+    this.version = setup('version', true)
+    this.config = setup('config', true)
+    this.keys = setup('keys', true, true)
+    this.blockstore = setup('blockstore', true)
+  }
+
+  exists (callback) {
+    this.version.exists(callback)
```
Comment (diasdavid) [inline src/index.js]: should it be through the `config`?

File: src/stores/blockstore.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
+'use strict'
+
+const Block = require('ipfs-block')
+const pull = require('pull-stream')
+const Lock = require('lock')
+const base32 = require('base32.js')
```
Comment (diasdavid) [inline src/stores/blockstore.js]: 👌🏽

File: src/stores/blockstore.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        push = cb
+      }
+
+      return {source, sink}
+    },
+
+    has (key, extension, cb) {
+      if (typeof extension === 'function') {
+        cb = extension
+        extension = undefined
```
Comment (diasdavid) [inline src/stores/blockstore.js]: In `get` we default to `data`, but in the remaining ops we don't.

```
const ext = extension === 'data' ? 'protobuf' : extension
```

Why is that? Any rationale?

File: src/stores/datastore.js [hunk 8]
```diff
@@ -1,103 +1,26 @@
 'use strict'
 
-const Lock = require('lock')
-const stream = require('stream')
-const bl = require('bl')
-const Block = require('ipfs-block')
+// TODO: This may end up being configurable
+// TODO: This should belong to the `fs` implementation
```
Comment (diasdavid) [inline src/stores/datastore.js]: ?

File: src/stores/datastore.js [hunk 9]
```diff
@@ -1,103 +1,26 @@
...
-    const path = multihashToPath(multihash, extension)
-    const through = stream.PassThrough()
-
-    lock(path, (release) => {
-      const ws = store.createWriteStream(path, release(cb))
-      through.pipe(ws)
-    })
-
-    return through
+module.exports = (store) => {
```
Comment (diasdavid) [inline src/stores/datastore.js]: datastore should be all levelDB, this current impl might confuse people

**Comment by diasdavid** [comment]
👍 for renaming things. Needs rebase, made some comments, then it is good to go :)

File: src/index.js (hunk 4 above)
Comment (dignifiedquire) [inline src/index.js]: why? this is not an assignment, just an array

File: src/index.js (hunk 5 above)
Comment (dignifiedquire) [inline src/index.js]: It is the same as before.

File: src/stores/blockstore.js (hunk 7 above)
Comment (dignifiedquire) [inline src/stores/blockstore.js]: It can actually be removed from get, as this is done in `multihashToPath`

File: src/stores/datastore.js (hunk 8 above)
Comment (dignifiedquire) [inline src/stores/datastore.js]: This is the code that was here before in datastore-legacy.js I did not touch it just rename it

File: src/stores/datastore.js (hunk 9 above)
Comment (dignifiedquire) [inline src/stores/datastore.js]: see above

Comment (dignifiedquire) [inline src/stores/datastore.js]: wait nevermind, this is wrong, going to remove this file. it makes no sense

File: .travis.yml (hunk 1 above)
Comment (dignifiedquire) [inline .travis.yml]: why? it is just additional time, and I have not seen a single thing working in 4 & 6 that could break in 5

File: src/index.js (hunk 4 above)
Comment (diasdavid) [inline src/index.js]: nvm :)

File: src/stores/blockstore.js (hunk 7 above)
Comment (diasdavid) [inline src/stores/blockstore.js]: - [ ] do it :D

**Comment by dignifiedquire** [comment]
@diasdavid ready :)

**Pull request merged**